reconcile/change_owners.py

```diff
@@ -705,36 +757,220 @@ def run(
+        # also in dry-run mode. this way change-owners can run in dry-run mode next
+        # to saas-file-owners and we can observe if bot integrations consider the MR
+        # self servicable or not.
+        labels = manage_conditional_label(
+            labels=labels,
+            condition=self_servicable,
+            true_label=SELF_SERVICEABLE,
+            false_label=NOT_SELF_SERVICEABLE,
+        )
+        if not dry_run:
+            labels = manage_conditional_label(
+                labels=labels,
+                condition=self_servicable and hold,
+                true_label=HOLD,
+            )
+            labels = manage_conditional_label(
+                labels=labels,
+                condition=self_servicable and approved,
+                true_label=APPROVED,
+            )
+            labels = manage_conditional_label(
+                labels=labels,
+                condition=self_servicable and not approved,
+                true_label=AWAITING_APPROVAL,
+            )
+        gl.set_labels_on_merge_request(gitlab_merge_request_id, labels)
+
+    except BaseException:
+        logging.error(traceback.format_exc())
+
+
+def manage_conditional_label(
+    labels: list[str],
+    condition: bool,
+    true_label: Optional[str] = None,
+    false_label: Optional[str] = None,
+) -> list[str]:
+    new_labels = labels.copy()
+    if condition:
+        if true_label and true_label not in labels:
+            new_labels.append(true_label)
+        if false_label and false_label in labels:
+            new_labels.remove(false_label)
+    else:
+        if true_label and true_label in labels:
+            new_labels.remove(true_label)
+        if false_label and false_label not in labels:
+            new_labels.append(false_label)
+    return new_labels
+
+
+@dataclass
+class Decision:
+
+    approve: bool = False
+    hold: bool = False
+
+
+@dataclass
+class ChangeDecision:
+
+    file: FileRef
+    diff: Diff
+    decision: Decision
+
+
+class DecisionCommand(Enum):
+    APPROVED = "/lgtm"
+    CANCEL_APPROVED = "/lgtm cancel"
+    HOLD = "/hold"
+    CANCEL_HOLD = "/hold cancel"
+
+
+def get_approver_decisions(comments: list[dict[str, str]]) -> dict[str, Decision]:
+    decisions_by_users: dict[str, Decision] = defaultdict(Decision)
+    for c in sorted(comments, key=lambda k: k["created_at"]):
+        commenter = c["username"]
+        for line in c.get("body", "").split("\n"):
+            if line == DecisionCommand.APPROVED.value:
+                decisions_by_users[commenter].approve = True
+            if line == DecisionCommand.CANCEL_APPROVED.value:
+                decisions_by_users[commenter].approve = False
+            if line == DecisionCommand.HOLD.value:
+                decisions_by_users[commenter].hold = True
+            if line == DecisionCommand.CANCEL_HOLD.value:
+                decisions_by_users[commenter].hold = False
```

Comment:
Use elif? Do you want to support multiple commands in one comment?

Reply:
`decisions_by_users` is a `defaultdict(Decision)`

Reply:
this replicates the current behaviour found in `saas-file-owners`. while `change-owners` runs in sort-of-a-sry-run mode next to `saas-file-owners` i would like their comment parsing behaviour to be the same, so we can compare results.

in general you have a point: most likely it does not make a lot of sense to have multiple (conflicting) commands in one comment.